permissions/src/androidLibMain/kotlin/AndroidPermissionsManager.kt

```diff
@@ -0,0 +1,131 @@
+package com.splendo.kaluga.permissions
+
+/*
+
+Copyright 2019 Splendo Consulting B.V. The Netherlands
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
+*/
+
+import android.content.Context
+import android.content.Intent
+import android.content.pm.PackageManager
+import androidx.core.content.ContextCompat
+import com.splendo.kaluga.base.ApplicationHolder
+import com.splendo.kaluga.log.debug
+import com.splendo.kaluga.log.error
+import com.splendo.kaluga.log.warn
+import kotlinx.coroutines.*
+import java.util.*
+import kotlin.concurrent.fixedRateTimer
+
+class AndroidPermissionsManager<P : Permission> constructor(private val context: Context = ApplicationHolder.applicationContext, private val permissionManager: PermissionManager<P>, private val permissions: Array<String> = emptyArray(), coroutineScope: CoroutineScope = permissionManager) : CoroutineScope by coroutineScope {
+
+    companion object {
+        const val TAG = "Permissions"
+
+        val lastPermission: MutableMap<String, Int> = mutableMapOf()
+        val waitingPermissions: MutableSet<String> = mutableSetOf()
+    }
+
+    private var timer: Timer? = null
+
+    fun requestPermissions() {
+        if (checkPermissionsDeclaration().isEmpty()) {
+            permissions.forEach {
```

Comment:
Thread safe?

Reply:
Moved the set into the launch, should ensure its on the main thread